receiver/prometheusreceiver/README.md

```diff
@@ -0,0 +1,452 @@
+
+
+### Metric Value Mapping
+ In OpenCensus, metrics value types can be either `int64` or `float64`, while in in Prometheus the value can be safety assume it's always `float64` based on the 
+[Prometheus Text Format Document](https://prometheus.io/docs/instrumenting/exposition_formats/#text-format-details) as quoted below:
+
+> value is a float represented as required by Go's ParseFloat() function. In addition to standard numerical values, Nan, +Inf, and -Inf are valid values representing not a number, positive infinity, and negative infinity, respectively.
+
+It will make sense for us to stick with this data type as much as possible across all metrics types
+
+
+### Counter
+Counter is one of the simplest data types we have in both Prometheus and Ocagent, it's mapping is as shown below
+
+For the following Prometheus Counters:
+```
+# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{method="post",code="200"} 1027
+http_requests_total{method="post",code="400"}    3
+```
+
+The corresponding Ocagent Metric will be:
+```go
+metrics := []*metricspb.Metric{
+  {
+    MetricDescriptor: &metricspb.MetricDescriptor{
+      Name:      "http_requests_total",
+      Type:      metricspb.MetricDescriptor_CUMULATIVE_DOUBLE,
+      LabelKeys: []*metricspb.LabelKey{{Key: "method"}, {Key: "code"}}},
+    Timeseries: []*metricspb.TimeSeries{
+      {
+        StartTimestamp: tsOc,
+        LabelValues:    []*metricspb.LabelValue{{Value: "post", HasValue: true}, {Value: "200", HasValue: true}},
+        Points: []*metricspb.Point{
+          {Timestamp: tsOc, Value: &metricspb.Point_DoubleValue{DoubleValue: 1027.0}},
+        },
+      },
+      {
+        StartTimestamp: tsOc,
+        LabelValues:    []*metricspb.LabelValue{{Value: "post", HasValue: false}, {Value: "400", HasValue: true}},
+        Points: []*metricspb.Point{
+          {Timestamp: tsOc, Value: &metricspb.Point_DoubleValue{DoubleValue: 3.0}},
+        },
+      },
+    },
+  },
+}
+```
+
+*Note: `tsOc` is a timestamp object representing the current ts*
+
+
+### Gauge
+For the following Prometheus Gauges:
+```
+# HELP gauge_test some test gauges.
+# TYPE gauge_test gague
+gauge_test{id="1",foo="bar"} 1.0
+gauge_test{id="2",foo=""}    2.0
+
+```
+
+The corresponding Ocagent Metric will be: 
+```go
+metrics := []*metricspb.Metric{
+  {
+    MetricDescriptor: &metricspb.MetricDescriptor{
+      Name:      "gauge_test",
+      Type:      metricspb.MetricDescriptor_GAUGE_DOUBLE,
+      LabelKeys: []*metricspb.LabelKey{{Key: "id"}, {Key: "foo"}}},
+    Timeseries: []*metricspb.TimeSeries{
+      {
+        StartTimestamp: tsOc,
+        LabelValues:    []*metricspb.LabelValue{{Value: "1", HasValue: true}, {Value: "bar", HasValue: true}},
+        Points: []*metricspb.Point{
+          {Timestamp: tsOc, Value: &metricspb.Point_DoubleValue{DoubleValue: 1.0}},
+        },
+      },
+      {
+        StartTimestamp: tsOc,
```

Comment:
Dose this logic also apply to Counters?

Reply:
Based on this comment, I have a feeling that are we expecting the counter to be a delta from OpenCensus in between two scrapes on the same target? If that's the case, the original implementation on Counter was correct, it did compute a delta with the previous value, however, it also computes delta on Gagues.  what about other data types, like histograms, the buckets also need to be a delta from the previous scrape?   @rghetia and @odeke-em  could you provide some specs on opencensus proto when a delta is needed, and when not, given that all sum/count/bucket_count values from prometheus are all cumulative.

Reply:
For counters it is delta between first (not previous) and current. Same for histogram sum/count/bucket_count. For Gauges it should be reported as is.

Reply:
Thank you for the reply. I am still a bit confused on the `delta`, since in Prometheus the values are recorded since the application started, the first value will be 0, observed or not.  I have also created a demo app with oclib and observed what was fed to the exporters, code as shown below
```
func main() {
	key, _ := tag.NewKey("foo")
	counter := stats.Int64("my-counter", "counter", "1")
	countView := &view.View{
		Measure:     counter,
		Name:        "simple-counter",
		Description: "Counter",
		Aggregation: view.Count(),
		TagKeys:     []tag.Key{key},
	}

	view.Register(countView)
	exp, _ := ocagent.NewExporter(ocagent.WithInsecure(), ocagent.WithServiceName("opencensus-demo"))
	defer exp.Stop()
	view.RegisterExporter(exp)

	// For demo purposes we'll use a very short view data reporting
	// period so that we can examine the stats send to our exporter, quickly.
	view.SetReportingPeriod(15 * time.Second)

	ctx, _ := tag.New(context.Background(), tag.Upsert(key, "bar"))
	for i := int64(0); ; i++ {
		stats.Record(ctx, counter.M(1))
		<-time.After(500 * time.Millisecond)
	}
}
```

from the ocagent side, the following metrics were seen from 2 reporting cycles. The first one is:
```
{
        "metric_descriptor": {
          "name": "simple-counter",
          "description": "Counter",
          "unit": "1",
          "type": 4,
          "label_keys": [
            {
              "key": "foo"
            }
          ]
        },
        "timeseries": [
          {
            "start_timestamp": {
              "seconds": 1560978301,
              "nanos": 352044000
            },
            "label_values": [
              {
                "value": "bar",
                "has_value": true
              }
            ],
            "points": [
              {
                "timestamp": {
                  "seconds": 1560978301,
                  "nanos": 352102000
                },
                "Value": {
                  "Int64Value": 30
                }
              }
            ]
          }
        ]
      }
```

and in the 2nd one:

```
{
        "metric_descriptor": {
          "name": "simple-counter",
          "description": "Counter",
          "unit": "1",
          "type": 4,
          "label_keys": [
            {
              "key": "foo"
            }
          ]
        },
        "timeseries": [
          {
            "start_timestamp": {
              "seconds": 1560978301,
              "nanos": 352044000
            },
            "label_values": [
              {
                "value": "bar",
                "has_value": true
              }
            ],
            "points": [
              {
                "timestamp": {
                  "seconds": 1560978316,
                  "nanos": 351525000
                },
                "Value": {
                  "Int64Value": 60
                }
              }
            ]
          }
        ]
      }
```

The first one has value 30, the 2nd one has 60. Basically, it's the value since `start_time`. Same observation found on histrograms too. This means the implementation from this PR is closed to the behaviors of OCLib (go). however, the start_time need to be adjust, it has to be the timestamp when the metrics is first recorded on the original system, however, this timestamp is not provided, the only available value is the first given timestamp when a prometheus target is scraped, and it needs to be cached for all following scrapes. Since targets are separated application, they can go down and restart at any given time. the prometheus receiver also need a way to detect such events and resetting the start_time stamp. we can compare any monotonically increasing metrics from a target, any decreasing of these values can be used as a signal.   

@rghetia  do you think the above description will be the correct behavior?

Reply:
What you have described is correct for the case 'a' below.

a. OcAgent is up and running. Demo app starts later and may or may not restart.
b. OcAgent restarts or starts after Demo App. 

In case 'b' there is no way to know when the Demo App started. May be it started few seconds ago or may be it started 10 days ago. If it started 10 days ago then it will show a big spike when rate/delta chart is plotted for the counter. For example, first value is 100000, second is 100010 after 1 minute, 100020 and so on. Delta chart would be flat line of 10 except for the first value which is 100000.

the other problem is that for backend like Stackdriver you cannot have timestamp of a point be the same as StartTime. Workaround there is to cache first timestamp after subtracting few seconds. 
Better solution to tackle both problem is to cache in the first value and timestamp. Do not report first value. Then report delta. You would still need to replace cached value and timestamp when  the value goes down (indicating that the demo app restarted).

Reply:
Nice, thanks for the clarification. would it be even better if we give an option to the user allow them to choose if they want to report the first value or not?

Reply:
Probably not. Due to very large first values, graph is impossible to use. I have experienced that myself with different receiver.

Reply:
Note that the same logic applies to histograms (count, sum, and buckets).

For summaries, it's a bit more complicated - count & sum can be handled similar to counters but percentiles are treated similar to gauges (since they capture an instant).